doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,243 @@
+
+As both of these authentication mechanisms include sending a password for
+authentication, they could also be used with application-specific passwords.
+As pure RADIUS provides only weak protection of the sent password, it might
+be even more desirable than in the case of traditional password authentication.
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+If the attribute for password in the user schema is not multi-valued, it must
+be modified to be. The user management must be extended to facilitate multiple
+passwords for a single user. This also includes password reset functionality -
+the user will be able to specify what password to reset. The user account
+status will display information for all its passwords.
+
+It is worth considering adding another attribute to the user schema, which will
+hold labels for all application-specific passwords, so that these labels can be
+displayed in the WebUI and used in the relevant commands in CLI, increasing
+the usability of these new functionality. The user will not need to remember
+that e.g. password no. 4 is for access to email on a smartphone.
+
+Other authentication mechanisms that also incorporate sending a password will
+also support application-specific passwords - namely otp (2FA), radius (RADIUS)
+and hardened (Hardened Password (by SPAKE or FAST)). 2FA will now allow to omit
+the OTP step if the client application does not support it and use only an
+application-specific password in such case.
+
+It is worth considering if:
+* the user will set application-specific passwords as usual (having to fulfil
+  password policies in place), and/or
+* they will be generated by FreeIPA (possibly only for one use, being an OTP).
+  It is usually done so by web services using application-specific passwords as
+  a fallback mechanism when 2FA is activated, but the client application does
+  not support it.
+
+### Permissions management for application-specific passwords
+
+The user will be able to configure permissions based on the password
+used to log into the account:
+* logging in using the primary password (or the master 2FA token)
+would grant full access to the account, including configuring all
+the other passwords as well as access to all services for which
+that account is used,
+* logging in with an application-specific password would allow
+access only to a specified subset of systems and services that
+password is allowed access to (for example, it could grant access
+only to log in to the user's mail server).
+
+By default, the password will have no permissions assigned, being inactive in
+practice, with the user then explicitly assigning permissions for it at his/her
+discretion. This may range from access to only one particular service up to
+either (TBD):
+* access rights on par with his primary password, or
+* access to all systems and services, but not to the account management.
```

Comment:
how would the user add permissions?
Users are not allowed to add ACIs so this need be hashed out in the proposal.

Reply:
I know, I envisage the permissions management for app-specific passwords to be logically separate from the (existing) general permissions management. I will make this clear in the design.